### PR TITLE
Create a just ESM version

### PR DESCRIPTION
**Pull request by wolframkriesing**
This PR tries to provide a solution for a pure-ESM package setup.
Solving #334 

This generates a pure ESM version which can be imported like `import create from "zustand/esm"`.
Not sure if this something of interest or if there is a much better and simpler solution.
Looking fwd to feedback.

In this file https://codeberg.org/wolframkriesing/bug-exploration-zustand-and-esm/src/branch/main/src/vanilla-esm.test.js I wrote tests that show this works. Might be a good idea putting them in here!?

Open issues
- [ ] build the middleware.ts

### CODE REVIEW

**Comment by dai-shi** [comment]
We've already done something more sophisticated in [jotai](https://github.com/pmndrs/jotai) and [valtio](https://github.com/pmndrs/valtio).

Would you add this in package.json?
```json
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./index.d.ts",
      "module": "./index.js",
      "default": "./index.cjs.js"
    },
    "./vanilla": {
      "types": "./vanilla.d.ts",
      "module": "./vanilla.module.js",
      "default": "./vanilla.js"
    },
    "./middleware": {
      "types": "./middleware.d.ts",
      "module": "./middleware.module.js",
      "default": "./middleware.js"
    },
    "./shallow": {
      "types": "./shallow.d.ts",
      "module": "./shallow.module.js",
      "default": "./shallow.js"
    }
  },
```

Oh, but you want to use "import", so it's different from "jotai" and "valtio"

```json
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./index.d.ts",
      "import": "./index.js",
      "default": "./index.cjs.js"
    },
    "./vanilla": {
      "types": "./vanilla.d.ts",
      "import": "./vanilla.module.js",
      "default": "./vanilla.js"
    },
    "./middleware": {
      "types": "./middleware.d.ts",
      "import": "./middleware.module.js",
      "default": "./middleware.js"
    },
    "./shallow": {
      "types": "./shallow.d.ts",
      "import": "./shallow.module.js",
      "default": "./shallow.js"
    }
  },
```

I don't know if this would cause issues with other environments.

File: rollup.config.js [hunk 1]
```diff
@@ -95,4 +95,11 @@ export default [
   createCommonJSConfig('src/shallow.ts', 'dist/shallow.js'),
   createCommonJSConfig('src/middleware.ts', 'dist/middleware.js'),
   createCommonJSConfig('src/vanilla.ts', 'dist/vanilla.js'),
+
+  // Create a purely ESM version, can be imported from `zustand/esm`
+  createESMConfig('src/vanilla.ts', 'dist/esm/vanilla.js'),
+  createESMConfig('src/shallow.ts', 'dist/esm/shallow.js'),
+  createESMConfig('src/index.ts', 'dist/esm/index.js'),
```
Comment (dai-shi) [inline rollup.config.js:102]:
```suggestion
  createESMConfig('src/vanilla.ts', 'dist/vanilla.module.js'),
  createESMConfig('src/shallow.ts', 'dist/shallow.module.js'),
  createESMConfig('src/index.ts', 'dist/index.module.js'),
```
Does this make sense? I'm not sure if "module" and "esm" is the same thing...
Anyway, my point is we want to make it consistent among zustand/jotai/valtio.
(I mean we could change the other two, if that's better.)

File: rollup.config.js [hunk 2]
```diff
@@ -95,4 +95,11 @@ export default [
...
+  // I dont get the one below to succeed with rollup :(
+  // createModernESMConfig('src/middleware.ts', 'dist/esm/middleware.js'),
```
Comment (dai-shi) [inline rollup.config.js:104]: 😕

**Comment by wolframkriesing** [comment]
I would import in an ESM package like this? `import create from "zustand/vanilla/import"`?

File: rollup.config.js (hunk 1 above)
Comment (wolframkriesing) [reply rollup.config.js:102]: looks good too. I have no strong opinion. I assume you have a broader overview and insights from users for what suits best.

**Comment by dai-shi** [comment]
If it's node14, it will read "exports" in package.json, so `import create from 'zustand/vanilla'` should work, I expect...

File: rollup.config.js (hunk 1 above)
Comment (dai-shi) [reply rollup.config.js:102]: As I understand, usually "module" and "esm" refer the same thing, but "mjs" does not.

**Comment by wolframkriesing** [comment]
Changed it to use your `exports` approach in the package.json.

File: rollup.config.js [hunk 3]
```diff
@@ -95,4 +95,11 @@ export default [
   createCommonJSConfig('src/shallow.ts', 'dist/shallow.js'),
   createCommonJSConfig('src/middleware.ts', 'dist/middleware.js'),
   createCommonJSConfig('src/vanilla.ts', 'dist/vanilla.js'),
+
+  // Create a purely ESM version, can be imported from `zustand/esm`
+  createESMConfig('src/vanilla.ts', 'dist/vanilla.module.js'),
+  createESMConfig('src/shallow.ts', 'dist/shallow.module.js'),
+  createESMConfig('src/index.ts', 'dist/index.module.js'),
```
Comment (dai-shi) [inline rollup.config.js:102]:
```suggestion
```

oops, we don't need this. it's `index.js`. this is inverted for historical reason.

**Comment by dai-shi** [comment]
Looks good. So what's remaining is:
- does it work in your use case?
- fix middleware.module.js
- how can we make sure if this doesn't break other envs...

File: package.json [hunk 4]
```diff
@@ -9,6 +9,29 @@
   "files": [
     "**"
   ],
+  "exports": {
+    "./package.json": "./package.json",
+    ".": {
+      "types": "./index.d.ts",
+      "import": "./index.js",
```
Comment (dai-shi) [inline package.json:16]:
```suggestion
      "module": "./index.js",
```
react doesn't support mjs, so we should only enable "import" for pure libs. let's make it back "module" for webpack5.

Comment (dai-shi) [reply package.json:16]: hm, I was confused with something else (= type="module"). Should be fine.

File: rollup.config.js (hunk 1 above)
Comment (wolframkriesing) [reply rollup.config.js]: same here. mjs is just the workaround of nodejs that exist(ed) while nodejs was figuring out how to move to real ESM

**Comment by wolframkriesing** [comment]
> how can we make sure if this doesn't break other envs...

seems to be the most challenging.
Using `exports` in package.json (afaiu) we limit import/require to what is listed in `exports`, if we forget one we break it for people.   
And afaik currently there are acouple of ways to import/require even just the `index.*`, e.g. you can import/require
`zustand`, `zustand/index`, `zustand/index.js`, `zustand/index.cjs`, `zustand/index.d.ts`, `zustand/index.iife.js`

Sounds like a couple of tests to ensure this wont break (until intended) might be a good idea.
For that I would be curious, what are the intended usages, starting with the `index.*`? All the above? Would a test ensure all the above keep working? If so I would start with a PR that provides tests for verifying all the current ways zustand can be imported/required. Wdyt?

**Comment by dai-shi** [comment]
we never had tests for `dist`. I'm not even sure if there's some best practice.
*.cjs, *.iife exist for historical reason, which I actually don't know.
Totally unsure if index.iife is used. index.cjs is used in some cases.

AFAIU, once all envs understand `"exports"`, it would be just good.
Meanwhile, we follow some basic conventions: defaults are cjs. `"main"` is cjs. `"module"` is esm.
People can probably use esm explicitly with `import ... from '*.module'`.

Now, jotai and valtio use "module" instead of "import" in "exports" to avoid dual package hazard.
Thinking of zustand, so far everything is pure, "import" would just work. But honestly, we don't want that because we may add some new features that depend on module variable.

Hmmm, I don't think my explanation is not well described.

----

Back to your original issue, type=module and `import create from 'valtio/vanilla'`.
- Does current build work for you? 9a4f7dd
- Does it fail if we modify "import" to "module" in "exports" in package.json? (It's probably same as just removing "import"s)

**Comment by dai-shi** [comment]
Hmm, I feel like I want to change the index.js:
- index.js -> index.module.js
- index.cjs -> index.js

This is a breaking change for users who use `import create from 'zustand/index.cjs'`.

**Comment by JeremyRH** [comment]
The IIFE build is usually used for getting directly from a CDN like unpkg. Unfortunately, we didn't configure it correctly and https://unpkg.com/zustand redirects to the CJS build. I guess it could also be used by some who use a more manual build process like gulp.
Too bad we cant see usage.

**Comment by dai-shi** [comment]
Thanks @JeremyRH . We are going towards ES module age in browser too, so I guess it's time to revisit bundles. It feels like CJS build works for most traditional use cases. My plan is to drop iife in the next minor and see what issues people report.

File: rollup.config.js (hunk 2 above)
Comment (rschristian) [reply rollup.config.js]: Curious what wasn't working? This not only works out of the box for me, but also gets rid of a `regenerator-runtime` dependency which is great. Huge improvement for us.

**Comment by dai-shi** [comment]
v3.4.1 comes with #343 and #350, which should almost do what is proposed in this PR.
Most discussions here are obsoleted, so let me close this and continue discussion in #334.

**Pull request closed without merging**